banderwagon/precomp_multiexp.go

```diff
@@ -1,135 +1,268 @@
+			return fmt.Errorf("serializing 8-bit table for %d-th point: %s", i, err)
+		}
 	}
 
-	for _, ltp := range pcl.inner {
-		for _, p := range ltp.matrix {
-			p.WriteUncompressedPoint(w)
+	// Serialize 16-bit tables.
+	if err := binary.Write(w, binary.LittleEndian, int64(len(pcl.inner16Bit))); err != nil {
+		return fmt.Errorf("serializing the number of points for 16-bit table: %s", err)
+	}
+	for i := range pcl.inner16Bit {
+		if err := pcl.inner16Bit[i].Serialize(w); err != nil {
+			return fmt.Errorf("serializing 16-bit table for %d-th point: %s", i, err)
 		}
 	}
 
 	return nil
 }
 
+// DeserializePrecomputedLagrange deserializes a PrecomputeLagrange.
+// See SerializePrecomputedLagrange() for the format description.
 func DeserializePrecomputedLagrange(reader io.Reader) (*PrecomputeLagrange, error) {
 	var pcl PrecomputeLagrange
 
 	var numPoints int64
-	err := binary.Read(reader, binary.LittleEndian, &numPoints)
-	if err != nil {
-		return nil, err
+	if err := binary.Read(reader, binary.LittleEndian, &numPoints); err != nil {
+		return nil, fmt.Errorf("deserializing the number of points: %s", err)
 	}
-	pcl.num_points = int(numPoints)
-	pcl.inner = make([]*LagrangeTablePoints, pcl.num_points)
+	pcl.numPoints = int(numPoints)
 
-	var rowLen int64
-	err = binary.Read(reader, binary.LittleEndian, &rowLen)
-	if err != nil {
-		return nil, err
+	// 8-bit table deserialization.
+	var table8BitCount int64
+	if err := binary.Read(reader, binary.LittleEndian, &table8BitCount); err != nil {
+		return nil, fmt.Errorf("deserializing the number of points for 8-bit table: %s", err)
+	}
+	pcl.inner8Bit = make([]*LagrangeTablePoints, table8BitCount)
+	for i := 0; i < int(table8BitCount); i++ {
+		pcl.inner8Bit[i] = &LagrangeTablePoints{}
+		pcl.inner8Bit[i].Deserialize(reader)
 	}
 
-	for i := 0; i < pcl.num_points; i++ {
-
-		pcl.inner[i] = new(LagrangeTablePoints)
-
-		pcl.inner[i].identity.Identity()
-
-		// Deserialize the matrix
-		pcl.inner[i].matrix = make([]bandersnatch.PointAffine, rowLen)
-		for j := int64(0); j < rowLen; j++ {
-			pcl.inner[i].matrix[j] = bandersnatch.ReadUncompressedPoint(reader)
-		}
+	// 16-bit table deserialization.
+	var table16BitCount int64
+	if err := binary.Read(reader, binary.LittleEndian, &table16BitCount); err != nil {
+		return nil, fmt.Errorf("deserializing the number of points for 16-bit table: %s", err)
+	}
+	pcl.inner16Bit = make([]*LagrangeTablePoints, table16BitCount)
+	for i := 0; i < int(table16BitCount); i++ {
+		pcl.inner16Bit[i] = &LagrangeTablePoints{}
+		pcl.inner16Bit[i].Deserialize(reader)
 	}
 
 	return &pcl, nil
 }
 
+// Commit computes the MSM of a set of evaluations.
 func (p *PrecomputeLagrange) Commit(evaluations []fr.Element) Element {
 	var result Element
 	result.Identity()
 
-	for i := 0; i < len(evaluations); i++ {
+	// We use p.inner16Bits for the first 5 group elements.
+	for i := 0; i < len(evaluations) && i < len(p.inner16Bit); i++ {
```

Comment:
I preferred using two separate loops for the different tables since their indexing in the corresponding table is slightly different. A single loop and an `if` inside won't avoid much repetition.